File: python/src/character_data_parser.py
import json
from character import Character, CharacterFinder
import re

character_finder = None


def probably_initialize_from_file_and_stuff(file_name=None):
    file_name = file_name or "strange_characters.json"
    with open(file_name, "r", encoding="utf-8") as file:
        data = json.load(file)

    all_characters = applesauce(data)
    global character_finder
    character_finder = CharacterFinder(all_characters)


def applesauce(data):
    result = []
    for character_data in data:
        character = Character(character_data["FirstName"], character_data["LastName"], character_data["IsMonster"])
        result.append(character)

    for character_data in data:
        # find nemesis
        if character_data["Nemesis"]:
            nemesis = find_character(result, character_data["Nemesis"])
            character = find_character(result, character_data["FirstName"])
            if character:
                character._nemesis = nemesis

        # add family
        if character_data.get("Children"):
            character = find_character(result, character_data["FirstName"])
            for child_name in character_data["Children"]:
                child = find_character(result, child_name)
                if child:
                    character._add_child(child)

    return result


def find_character(characters, first_name):
    found = [c for c in characters if c.firstName == first_name]
    if found:
        return found[0]
    return None


def evaluate_path(path):
    character = None
    has_family_name = False
    character_name = ""
    family_name = ""
    temp_path_without_curly_braces = ""
    curly_braces = ""
    structure_list = path.split("/")

    for i in range(len(structure_list) - 1, -1, -1):
        local_name = ""
        local_name_without_curly_braces = ""
        family_local_name_list = structure_list[i].split(":")

        if len(family_local_name_list) == 2:
            if not has_family_name:
                family_name = family_local_name_list[0]
                has_family_name = True
            family_name = family_local_name_list[0]
            local_name = family_local_name_list[1]
        elif len(family_local_name_list) == 1:
            local_name = family_local_name_list[0]

        if i == len(structure_list) - 1:
            character_name = local_name

        local_name_without_curly_braces = re.sub(r'{[^{]*?}', '', local_name)
        matches = re.findall(r'(.*){([^}]*)}', local_name)

        for match in matches:
            groups = match
            curly_braces = groups[1]
            character_name = groups[0]

        temp_path_without_curly_braces = f"/{local_name_without_curly_braces}" + temp_path_without_curly_braces

    if not has_family_name:
        character = character_finder.find_by_first_name(character_name)
        if character and curly_braces and curly_braces == "Nemesis":
            character = character_finder.find_by_first_name(character_name)
            if character and curly_braces and curly_braces == "Nemesis":
                return character._nemesis
            return character

    filtered_characters = character_finder.find_family_by_last_name(family_name)
    if filtered_characters:
        names = temp_path_without_curly_braces.split("/")
        first_name = names[-1]
        found = [c for c in filtered_characters if c.firstName == first_name]
        if found:
            character = found[0]
        else:
            character = None

        if character and curly_braces and curly_braces == "Nemesis":
            return character._nemesis

    return character

File: cpp/src/CharacterDataParser.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <regex>

#include "CharacterDataParser.hpp"
#include "CharacterFinder.hpp"
#include "nlohmann/json.hpp"

using json = nlohmann::json;

namespace {
    std::vector<std::string> split(std::string s, char delimiter)
    {
        std::vector<std::string> tokens;
        std::string token;
        std::istringstream tokenStream(s);
        while (std::getline(tokenStream, token, delimiter))
        {
            if (!token.empty())
                tokens.push_back(token);
        }
        return tokens;
    }
}

namespace Characters {
    std::vector<std::shared_ptr<Character>> CharacterDataParser::allCharacters;
    
    void CharacterDataParser::probably_initializeFromFile_andstuff(const std::string &filename) {
        allCharacters.clear();

        std::ifstream data_stream(filename, std::ifstream::in);
        if (!data_stream.is_open()) {
            std::cout << "Error opening file:" << filename << '\n';
            std::cout << "Error: " << strerror(errno) << '\n';
            return;
        }
        auto data = json::parse(data_stream);
        applesauce(data);
    }

    void CharacterDataParser::applesauce(nlohmann::basic_json<> data) {
        for (auto characterData : data)
        {
            auto character = new Character(
                characterData["FirstName"].get<std::string>(),
                characterData["LastName"].is_string() ?
                    characterData["LastName"].get<std::string>() : std::string(),
                characterData["IsMonster"].get<bool>()
            );
            allCharacters.emplace_back(character);
        }

        for (auto characterData : data)
        {
            if (!characterData["Nemesis"].is_null())
            {
                auto nemesis = findCharacter(allCharacters, characterData["Nemesis"].get<std::string>());
                auto character = findCharacter(allCharacters, characterData["FirstName"].get<std::string>());

                if (character.has_value())
                    (*character)->SetNemesis(nemesis);
            }

            if (!characterData["Children"].is_null())
            {
                auto character = findCharacter(allCharacters, characterData["FirstName"].get<std::string>());
                for (auto childName : characterData["Children"])
                {
                    auto child = findCharacter(allCharacters, childName.get<std::string>());
                    if (child.has_value())
                        (*character)->AddChild(*child);
                }
            }
        }
    }

    std::optional<Character *> CharacterDataParser::evaluatePath(const std::string &path) {
        std::optional<Character*> character = std::nullopt;
        CharacterFinder characterFinder(allCharacters);

        auto hasFamilyName = false;
        auto characterName = std::string();
        auto familyName = std::string();
        auto tempPathWithoutCurlyBraces = std::string();
        auto curlyBraces = std::string();

        auto structureList = split(path, '/');

        for (int i = structureList.size() - 1; i >= 0; --i)
        {
            auto localName = std::string();
            auto localNameWithoutCurlyBraces = std::string();
            auto familyLocalNameList = split(structureList[i], ':');
            if (familyLocalNameList.size() == 2)
            {
                if (!hasFamilyName)
                {
                    familyName = familyLocalNameList[0];
                    hasFamilyName = true;
                }

                familyName = familyLocalNameList[0];
                localName = familyLocalNameList[1];
            }
            else if (familyLocalNameList.size() == 1)
            {
                localName = familyLocalNameList[0];
            }

            if (i == structureList.size() - 1)
            {
                characterName = localName;
            }

            localNameWithoutCurlyBraces = std::regex_replace(localName, std::regex("\\{[^{]*?\\}"), "");

            auto matches = std::smatch { };

            if (std::regex_match(localName, matches, std::regex("(.*)\\{([^{]*)\\}")))
            {
                curlyBraces = matches[2];
                characterName = matches[1];
            }

            tempPathWithoutCurlyBraces = tempPathWithoutCurlyBraces.insert(0, "/" + localNameWithoutCurlyBraces);
        }

        if (!hasFamilyName)
        {
            character = findCharacter(allCharacters, characterName);
            if (curlyBraces == "Nemesis")
            {
                return (*character)->Nemesis;
            }
            return character;
        }

        auto filteredCharacters = characterFinder.FindFamilyByLastName(familyName);
        if (!filteredCharacters.empty())
        {
            auto firstName = std::string_view(tempPathWithoutCurlyBraces).substr(
                std::string_view(tempPathWithoutCurlyBraces).find_last_of("/")+1
            );
            auto c = std::find_if(filteredCharacters.begin(), filteredCharacters.end(),
                                  [&firstName](const auto& character) { return character->FirstName == firstName; });
            if (c != filteredCharacters.end()) {
                character = *c;
            } else {
                character = std::nullopt;
            }
        
            if (character != nullptr && curlyBraces == "Nemesis")
            {
                return (*character)->Nemesis;
            }
        }

        return character;
    }

    std::optional<Character*> CharacterDataParser::findCharacter(std::vector<std::shared_ptr<Character>> characters, std::string_view firstName) {
        auto c = std::find_if(characters.begin(), characters.end(),
                [&firstName](const auto& character) { return character->FirstName == firstName; });

        if (c != characters.end())
            return c->get();
        return std::nullopt;
    }
}
